### PR TITLE
Swizzle expression

### LINKED ISSUE #681: Swizzle expression in IR
The reason we don't have `Expression::Swizzle` today is because it's not required for anything. However, without swizzling we end up naively generating code like this:
```cpp
metal::float3(_expr23.x, _expr23.y, _expr23.z)
```
Given that both our input languages and output languages support swizzling (on vectors, at least), it seems more trouble than it's worth (i.e. if we are trying to keep the IR minimal) to avoid them in the IR. Let's have them?

### PR DESCRIPTION
**Pull request by kvark**
Closes #681
As you can see, the produced output is much better now.

@grovesNL there is one controversial thing here - removal of the invalid access test in WGSL. We were generally trying to have a single place where inputs are validated. So WGSL code only has to validate the state that the IR validation doesn't cover. The access expressions are fully covered now, so technically WGSL should leave it to IR for validation.
However, the IR validation errors are nearly not as nice, and this is a general problem we have that needs more thought. We can consider #358 to be this issue.

### CODE REVIEW

**Review by grovesNL — APPROVED** [review]
Looks good to me! Yeah I think we'll be able to restore the validation once we pass span information through to the validator

File: src/front/wgsl/mod.rs [hunk 1]
```diff
@@ -379,55 +379,49 @@ impl<'a> ExpressionContext<'a, '_, '_> {
...
-        name_span: Range<usize>,
-    ) -> Result<crate::Expression, Error> {
+    fn extract_impl(name: &str, name_span: Range<usize>) -> Result<u32, Error> {
         let ch = name
             .chars()
             .next()
             .ok_or_else(|| Error::BadAccessor(name_span.clone()))?;
-        let index = Self::letter_pos(ch);
-        if index >= base_size as u32 {
-            return Err(Error::BadAccessor(name_span));
```
Comment (jimblandy) [inline src/front/wgsl/mod.rs:410]: So is the idea that the expression validator now checks that the accessors make sense for the input vector type, so we don't need to check for that here?

**Review by jimblandy — COMMENTED** [review]
Looks good - just some suggestions.

File: src/lib.rs [hunk 2]
```diff
@@ -679,6 +679,21 @@ pub enum ImageQuery {
     NumSamples,
 }
 
+/// Component selection for a vector swizzle.
+#[derive(Clone, Copy, Debug, PartialEq)]
+#[cfg_attr(feature = "serialize", derive(Serialize))]
+#[cfg_attr(feature = "deserialize", derive(Deserialize))]
+pub enum SwizzleComponent {
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/lib.rs:686]: Good idea! We do this for `VectorSize`, and we should definitely do it here.

File: src/front/wgsl/mod.rs (hunk 1 above)
Comment (kvark) [reply src/front/wgsl/mod.rs:410]: right, we don't want to implement the same check in multiple places if this can be avoided

**Comment by kvark** [comment]
Thanks @jimblandy ! Everything should be addressed now with #750

**Pull request merged**